Content.Server/GameObjects/Components/Power/ApcNetComponents/PowerReceiverComponent.cs

```diff
@@ -0,0 +1,211 @@
+        ///     Amount of charge this needs from an APC per second to function.
+        /// </summary>
+        [ViewVariables]
+        public float Load { get => _load; set => SetLoad(value); }
+        private float _load;
+
+        /// <summary>
+        ///     The fraction of APC charge below which this shuts off and stops using power.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public float PowerShutoffFraction { get => _powerShutoffFraction; set => SetPowerShutoffFraction(value); }
+        private float _powerShutoffFraction;
+
+        /// <summary>
+        ///     When true, causes this to appear powered even if not receiving power from an Apc.
+        /// </summary>
+        [ViewVariables(VVAccess.ReadWrite)]
+        public bool NeedsPower { get => _needsPower; set => SetNeedsPower(value); }
+        private bool _needsPower;
+
+        /// <summary>
+        ///     When true, causes this to never appear powered.
+        /// </summary>
+        public bool PowerDisabled { get => _powerDisabled; set => SetPowerDisabled(value); }
+        private bool _powerDisabled;
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _powerReceptionRange, "powerReceptionRange", 3);
+            serializer.DataField(ref _load, "powerLoad", 5);
+            serializer.DataField(ref _powerShutoffFraction, "powerShutoffFraction", 0.3f);
+            serializer.DataField(ref _needsPower, "needsPower", true);
+            serializer.DataField(ref _powerDisabled, "powerDisabled", false);
+        }
+
+        protected override void Startup()
+        {
+            base.Startup();
+            if (NeedsProvider)
+            {
+                TryFindAndSetProvider();
+            }
+        }
+
+        public override void OnRemove()
+        {
+            _provider.RemoveReceiver(this);
+            base.OnRemove();
+        }
+
+        public void TryFindAndSetProvider()
+        {
+            if (TryFindAvailableProvider(out var provider))
+            {
+                Provider = provider;
+            }
+        }
+
+        private bool TryFindAvailableProvider(out IPowerProvider foundProvider)
+        {
+            var nearbyEntities = IoCManager.Resolve<IServerEntityManager>()
+                .GetEntitiesInRange(Owner, PowerReceptionRange);
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            foreach (var entity in nearbyEntities)
+            {
+                if (entity.TryGetComponent<PowerProviderComponent>(out var provider))
+                {
+                    if (provider.Connectable)
+                    {
+                        var distanceToProvider = provider.Owner.Transform.GridPosition.Distance(mapManager, Owner.Transform.GridPosition);
+                        if (distanceToProvider < Math.Min(PowerReceptionRange, provider.PowerTransferRange))
```

Comment:
Is there a point to the `Min` here, since the entity is guaranteed to be in `PowerReceptionRange` by the query?

Reply:
The GetEntitiesInRange() query returns entities in a square range, so entities in the corner will be outside a circle of r = PowerReceptionRange.